### ISSUE TITLE
(inactive) PhosMap

### ISSUE THREAD

**Issue by ecnuzdd** [Bioconductor/Contributions#935]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/ecnuzdd/PhosMap

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For help with submitting your package, please subscribe and post questions
to the [bioc-devel][4] mailing list.

[1]: https://bioconductor.org/developers/package-guidelines/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/

**Comment by mtmorgan** [comment]
This package does not have a [vignette][1], and does not re-use standard Bioconductor [classes and methods][2] for interoperability, and is therefore not suitable for Bioconductor.

[1]: https://bioconductor.org/packages/release/bioc/html/BiocStyle.html
[2]: https://bioconductor.org/developers/how-to/commonMethodsAndClasses/

**Comment by ecnuzdd** [comment]
Hi, I have updated PhosMap, please help me reopen the issue to make it work.

**Comment by mtmorgan** [comment]
Your package can depend only on CRAN or Bioconductor packages.

**Comment by ecnuzdd** [comment]
Does  Bioconducor  not support non-CRAN or non-Bioconductor packages? The rmotifx and ksea packages are not located on CRAN and Bioconductor. In DESCRIPTION file, I wrote that: Additional_repositories: https://github.com/omarwagih/rmotifx/,https://github.com/evocellnet/ksea/

**Comment by mtmorgan** [comment]
Packages must be on CRAN or Bioconductor, since these have at least minimal quality control and certainty of availability.

**Comment by ecnuzdd** [comment]
In proteome, kinase-substrate enrichment analysis (KSEA) and Motif enrichment analysis (rmotifx) are two general methods. For this case, how do I use the two packages? In DESCRIPTION file, if I modified them to "Suggests" labels, does it work? it is successful while executing 'R CMD build'.

**Comment by mtmorgan** [comment]
It's successful from R CMD build because you have the packages installed. The build system will not install them, and they cannot be in the Depends:, Imports:, or Suggests: fields. If these packages are central to the functionality of your own package, then it means that your package cannot be accepted to Bioconductor.

**Comment by ecnuzdd** [comment]
Hi, reviewer,
I have updated my package and passed "R CMD check" and "R CMD build" in local development machine (Windows). There were no ERRORS and WARNINGs happening. I don't konw why it happend that "Quitting from lines 254-323 (introduction_of_PhosMap.Rmd) 
Error: processing vignette 'introduction_of_PhosMap.Rmd' failed with diagnostics:
numbers of columns of arguments do not match
Execution halted".
I have also executed these cmd located in lines 254-323 by pefroming a single steps, it looks anything is OK. Please help me find out why and the I will figure out it.

**Comment by mtmorgan** [comment]
I tried to build your package but apparently you have commited an invalid version of the file data/BRAFi.rda
```
PhosMap master$ bioc-devel
> load("data/BRAFi.rda")
Error in load("data/BRAFi.rda") :
  bad restore file magic number (file may be corrupted) -- no data loaded
In addition: Warning message:
file 'BRAFi.rda' has magic number 'versi'
  Use of save versions prior to 2 is deprecated
```

**Comment by ecnuzdd** [comment]
Thanks for your operations. 
In this version, I have commented  lines 254-323 in "introduction_of_PhosMap.Rmd", so you can see the error. I'm trying to solve it and then submit a new version. Currently, I don't find a correct method.
I also have another question, does Bioconductor allow for embedded library files over 100 MB?

**Comment by mtmorgan** [comment]
'embedded library file'? If you mean compiled (e.g., C) code, then that is not the correct solution.

Please do not comment out lines that do not work, as this makes it difficult to know which lines do not work!!!

**Comment by ecnuzdd** [comment]
The 'embedded library file' means that references files used for analysis, sucha as protein fasta file from NCBI, the collected kinase-substrate regulation relations and constructed background file for motif enrichment analysis. The existence of files are to help users start perform analysis rapidly and expediently, reduce time cost of file collection and background construction. All these files are put into 'inst/extdata' directory for use.

**Comment by mtmorgan** [comment]
Large data files should not be included in the package. If they are easily available from online sources then I suggest writing a small helper function that uses [BiocFileCache][1] to retrieve the files (the first time) and read them from a local cache (first and subsequent times). If the data require curation (by you) before being useful, then they should be added to [ExperimentHub][2].

[1]: https://bioconductor.org/packages/BiocFileCache
[2]: https://bioconductor.org/packages/devel/bioc/vignettes/ExperimentHub/inst/doc/CreateAnExperimentHubPackage.html

**Comment by mtmorgan** [comment]
Will you be updating your package, or shall we mark this as 'inactive'?

**Comment by mtmorgan** [comment]
Please increment the version number (to 0.99.14) to trigger a build. Please follow this issues (https://github.com/Bioconductor/Contributions/issues/935) for further updates

**Comment by ecnuzdd** [comment]
Thanks for your quick reply. 


I will update version number in description to trigger a build.




Best regards!
Dongdong Zhan
The Institute of Biomedical Sciences
East China Normal University
500 Dongchuan Road
Shanghai 200241, China
On 5/9/2019 18:17，Martin Morgan<notifications@github.com> wrote：

Please increment the version number (to 0.99.14) to trigger a build. Please follow this issues (#935) for further updates

**Comment by ecnuzdd** [comment]
Hello, I updeted  version number in DESCRIPTION file and pushed a request via webhook trigger. But the operation did not work , because Github showed a prompt: "We couldn’t deliver this payload: Service Timeout" . 
Meanwhile,  I do not receive a notice email. Please help me check whether my submission works.
Many thanks.

**Comment by ecnuzdd** [comment]
Hello, I have seriously check the build report. All compiling is ok except that there is only one warning when running "R CMD check". 

The warnning prompts "R CMD check exceeded 10 min requirement", this is because PhosMap needs to load reference library when applying it to process Phosphoproteomics data. In our examples, we put these large files  on a FTP server, once the example is tested, it first grasp reference data from FTP. The process would take a few minutes, which finally result in the warning from package builder of Biocoductor .

So,  could the warning be ignored? If not, please give me some suggestion about fixing it.

**Comment by ecnuzdd** [comment]
@bioc-issue-bot Hello, I have seriously check the build report. All compiling is ok except that there is only one warning when running "R CMD check".

The warnning prompts "R CMD check exceeded 10 min requirement", this is because PhosMap needs to load reference library when applying it to process Phosphoproteomics data. In our examples, we put these large files on a FTP server, once the example is tested, it first grasp reference data from FTP. The process would take a few minutes, which finally result in the warning from package builder of Biocoductor .

So, could the warning be ignored? If not, please give me some suggestion about fixing it.

**Comment by mtmorgan** [comment]
Have you thought of using [BiocFileCache][1] to download the data only once? Or better, to make your data available on [AnnotationHub][2] or the related ExperimentHub so that it is there even after you lose interested, and is automatically cached locally?

[1]: https://bioconductor.org/packages/BiocFileCache
[2]: https://bioconductor.org/packages/AnnotationHub

**Comment by ecnuzdd** [comment]
@mtmorgan Thanks for your reply. Maybe I don't describle the question clearly.

My question is that  "R CMD check"  would test examples written in each Rscript, the smooth execution of these partial examples  is dependent on some libraries with large sizes (> 200 MB), whose loading time will result in the warnning "R CMD check exceeded 10 min requirement". These loading process does not matter where the library file exists or no matter how to cache them. 

Of course, when users perform PhosMap in practice, all dependence libraries will cache locally for reusing.

I am sure that all Rscript is OK, the only warnning at the moment is "R CMD check" exceeds time while it does not come from performability of Rscript.

**Comment by mtmorgan** [comment]
Revise the evaluated code so that it remains comprehensive but evaluates within the specified time limit. It is not satisfactory to assume that the R scripts are ok -- they may be now, but changes in your package or packages that your package depends on can break them in the future. Instead of running code on very large data, create an example data set that is enough to illustrate the analysis and to test your code.

The bottom line is that your package must build and check within specified time limits.

**Comment by mtmorgan** [comment]
#  6 June, 2019

Sorry for my very slow review.

Generally, I'm concerned about the reliance on an ftp site for data in your package; past experience has been that the author does not maintain these sites in the long term, meaning that users cannot rely on package functionality. I'm also concerned about the efficiency of the code, especially the absence of vectorization and use of 'copy-and-append' as indicated below. Finally many of your examples do not evaluate because they have `if(FALSE){}`; this renders the examples useless and the examples should be revised so that they are fully evaluated, with only rare exceptions. Likely this means using small sample data included with the package, or data made available in ExperimentHub.

DESCRIPTION

- good

NAMESPACE

- Looks like this is a combination of roxygen and 'hand' generated; this is very confusing for those looking at your code. Use one style or the other.

NEWS

- Make sure your formatting can be parsed by utils::news()

vignettes

- All data used in your vignette must either be in your package, or in an ExperimentHub resource; it is not acceptable to download data from a private server or github

- wrap lines carefully in code chunks so that they do not spill over page bounds on output

- use `fl <- tempfile(); dir.create(fl)` rather than `getwd()`, `~/`, or other directories in the user's file system.

- Provide more narrative text to help the user understand the inputs and outputs of your code chunks.

- Avoid large code chunks without intervening narrative text.

R (specific examples identified in the comments, but applicable throughout)

- `analysis_deps_anova.R:75`; `extract_psites_score.R:90` and many other places: avoid 'copy-and-append' pattern; instead, vectorize (best) or 'pre-allocate and fill', see http://bioconductor.org/developers/how-to/efficient-code/

- `analysis_deps_anova.R:33` use `@import` or `@importFrom` to make functions from other packages available in your package; avoid using `requireNamespace()`.

- use `stop()`, `warning()`, `message()` to signal error or informative messages to the user; use `cat()` only to display an object in a `print()` or `show()` method.
  ```
  if(group_nlevels < 2){
      cat('\n', 'Not found pairwise comparison.', '\n')
      stop('')
  }
  ```
  should be
  ```
  if(group_nlevels < 2)
      stop('Not found pairwise comparison.')
  ```

- `load_data_with_ftp.R:22` use BiocFileCache to manage data downloaded by the user

- `get_aligned_seq_for_mea.R:90` this code chunk looks like it could be vectorized, and could be re-written using efficient the effiicent Biostrings pacakge.

man

- Wrap man page examples, including comment strings to 80 characters so that they display clearly in a standard console

- Do not try to avoid evaluating example code with `if (FALSE) {}`, `\dontrun{}` or `\donttest{}`; `if (FALSE){}` is never appropriate; the latter are appropriate when a limited section of the example cannot be run by the user interactively (e.g., over-writing files) or during check (e.g., requiring user interaction).

**Comment by ecnuzdd** [comment]
Hi Martin，


Thanks for your careful review and constructive suggestion.


I will comply with these suggestion to perform point-by-point updating.


As for your concern about the reliance on an ftp site for data, I want to state that I specially apply for the ftp site from our network center to store demo data and library data of PhosMap and provide users with more smooth experience. I promise we will continuously maintain it and guarantee it works. 


Your suggestion about R code is helpful for us, I will further standardize them.


The if(FALSE){} is an inappropriate format, I will replace them with \donttest{} for check.


Best regards!


Dongdong Zhan
The Institute of Biomedical Sciences
East China Normal University
500 Dongchuan Road
Shanghai 200241, China
On 6/7/2019 05:43，Martin Morgan<notifications@github.com> wrote：
6 June, 2019

Sorry for my very slow review.

Generally, I'm concerned about the reliance on an ftp site for data in your package; past experience has been that the author does not maintain these sites in the long term, meaning that users cannot rely on package functionality. I'm also concerned about the efficiency of the code, especially the absence of vectorization and use of 'copy-and-append' as indicated below. Finally many of your examples do not evaluate because they have if(FALSE){}; this renders the examples useless and the examples should be revised so that they are fully evaluated, with only rare exceptions. Likely this means using small sample data included with the package, or data made available in ExperimentHub.

DESCRIPTION

good

NAMESPACE

Looks like this is a combination of roxygen and 'hand' generated; this is very confusing for those looking at your code. Use one style or the other.

NEWS

Make sure your formatting can be parsed by utils::news()

vignettes

All data used in your vignette must either be in your package, or in an ExperimentHub resource; it is not acceptable to download data from a private server or github

wrap lines carefully in code chunks so that they do not spill over page bounds on output

use fl <- tempfile(); dir.create(fl) rather than getwd(), ~/, or other directories in the user's file system.

Provide more narrative text to help the user understand the inputs and outputs of your code chunks.

Avoid large code chunks without intervening narrative text.

R (specific examples identified in the comments, but applicable throughout)

analysis_deps_anova.R:75; extract_psites_score.R:90 and many other places: avoid 'copy-and-append' pattern; instead, vectorize (best) or 'pre-allocate and fill', see http://bioconductor.org/developers/how-to/efficient-code/

analysis_deps_anova.R:33 use @import or @importFrom to make functions from other packages available in your package; avoid using requireNamespace().

use stop(), warning(), message() to signal error or informative messages to the user; use cat() only to display an object in a print() or show() method.

if(group_nlevels < 2){
    cat('\n', 'Not found pairwise comparison.', '\n')
    stop('')
}


should be

if(group_nlevels < 2)
    stop('Not found pairwise comparison.')


load_data_with_ftp.R:22 use BiocFileCache to manage data downloaded by the user

get_aligned_seq_for_mea.R:90 this code chunk looks like it could be vectorized, and could be re-written using efficient the effiicent Biostrings pacakge.

man

Wrap man page examples, including comment strings to 80 characters so that they display clearly in a standard console

Do not try to avoid evaluating example code with if (FALSE) {}, \dontrun{} or \donttest{}; if (FALSE){} is never appropriate; the latter are appropriate when a limited section of the example cannot be run by the user interactively (e.g., over-writing files) or during check (e.g., requiring user interaction).

**Comment by mtmorgan** [comment]
If your package is ready for further review, please be sure to increment the version to trigger a build.

**Comment by ecnuzdd** [comment]
Hi Martin，

Thanks for your kind reminders.

We are updating PhosMap. Its new version will be coming soon.

Best regards!

Dongdong

**Comment by mtmorgan** [comment]
I'll mark this package as inactive. Comment here when you are ready to resume the review process.
